csharp/src/Ice/OutgoingFrame.cs

```diff
@@ -309,29 +324,34 @@ internal virtual void Finish()
                             istr.Skip(entrySize);
                             if (!ContainsKey(key))
                             {
-                                Data.Add(_defaultBinaryContext.Slice(startPos, istr.Pos - startPos));
+                                Data.Add(defaultBinaryContext.Slice(startPos, istr.Pos - startPos));
                                 AddKey(key);
                             }
                         }
                     }
 
-                    ostr.RewriteFixedLengthSize20(_binaryContextKeys.Count, _payloadEnd, 2);
+                    ostr.RewriteFixedLengthSize20(_binaryContextKeys.Count, PayloadEnd, 2);
                 }
                 else
                 {
                     Debug.Assert(_binaryContextKeys == null);
-                    if (_defaultBinaryContext.Count > 0) // only when forwarding an ice2 request or response
-                    {
-                        Debug.Assert(Data[^1].Array == _defaultBinaryContext.Array);
 
+                    // Only when forwarding an ice2 request or response
+                    if (defaultBinaryContext.Count > 0 && Data[^1].Array == defaultBinaryContext.Array)
```

Comment:
Shouldn't we also copy the default binary context here instead of just expanding the last segment?

Reply:
Here we expand the last segment to include the default binary context.

As per the comment, the default binary context is non-empty only when "forwarding" a frame. In this forwarding situation, the payload segment and the default binary context share the same segment/array. This block is an optimization for the situation where no Context or binary context was written / needs to be written: in this case, rather than copying the default binary context, we can just expand the payload segment to include the default binary context.